### DISCUSSION TITLE
Error 18 in Adobe Reader and Acrobat

### DISCUSSION THREAD

**Discussion by djoltes** [pymupdf/PyMuPDF#2590] [Looking for help] [answered]
I have my redaction code working well for PII, signatures, etc. but have encountered one problem. Redacted files open perfectly in Foxit, Chrome, and other readers but we get page errors from Adobe products. I've been digging around to find what error 18 is when loading a page, but haven't found a definitive listing of Adobe error codes. 

![image](https://github.com/pymupdf/PyMuPDF/assets/11085091/d2b5bca5-8e6b-4a3f-81ee-03f79c8d89b0)

Other pages in the docs render fine, but some throw this error. It doesn't seem to be a Windows system error (18 is ERROR_NO_MORE_FILES) and all the Google hits I get for that error from Adobe are from pre-2010.

**Comment by JorjMcKie** [comment]
Please provide an example file with this behaviour - and also your PyMuPDF version.
I forgot:
Please also give us a code snippet that does the redaction leading to this result.

**Comment by djoltes** [comment]
Here's a 2-page file that exhibits the behaviour.  Version is 1.22.5, so the latest AFAIK
[new_test_doc_redacted.pdf](https://github.com/pymupdf/PyMuPDF/files/12335140/new_test_doc_redacted.pdf)

For the redaction, there are several blocks (one for signatures, another for detected PII, etc.) but they look like this.

Signatures:
 ```
                    geo = item["Geometry"]
                    box = geo["BoundingBox"]
        
                    x0 = (box["Left"] / IMG_OFFSET) * iWidth # left side
                    y0 = (box["Top"] / IMG_OFFSET) * iHeight
                    height = (box["Height"] / IMG_OFFSET) * iHeight 
                    width = (box["Width"] / IMG_OFFSET) * iWidth
                    x1 = x0 + width # computed width
                    y1 = y0 + height # computed height
                    print("Signature at: {a} {b}, RH: {c}, {d}".format(a=x0, b=y0, c=x1, d=y1))
                    # create a rectangle based on the above
                    r = fitz.Rect(x0, y0, x1, y1)
                    # Now add our redaction annotation based on the rectangle we've built
                    annot = page.add_redact_annot(r, fill = blue, text=REDACT_STR, text_color = blue) 
```
Another set...ignore bad indents from the paste. These handle, respectively, PII detected by a service and ITIN numbers identified via a regex.
```
            if piiList is not None:
                    for data in piiList:
                        # print("email: {a}".format(a=data))
                        areas = page.search_for(data)
                        print("PII entity: {a}".format(a = data.encode('utf-8')))  
                        # logging.info("PII entity: {a}".format(a = data.encode('utf-8')))          
                        # draw outline over sensitive data areas based on coordinates on the page
                        [page.add_redact_annot(area, fill = (0, 0, 0), text=REDACT_STR) for area in areas]                
                            
            itinNums = getITIN(pageContent)
            if itinNums:
                for itinNum in itinNums:
                    logging.info("ITIN: {a}".format(a=itinNum))
                    print("ITIN: {a}".format(a=itinNum))
                    areas = page.search_for(itinNum)
                
                    # draw outline over sensitive data contained in the locations identified earlier
                    [page.add_redact_annot(area, fill = (0, 0, 0), text=REDACT_STR) for area in areas]      
```

**Reply by JorjMcKie** [reply]
Thanks for the file!
Sorry to be so incremental in my requests: Can you please also share the **original** file? I need to reproduce the what happens to see whether we have a deeper problem.

**Reply by djoltes** [reply]
Oh, no worries -- here it is. I probably should have attached it to the last reply! 
[new_test_doc.pdf](https://github.com/pymupdf/PyMuPDF/files/12360102/new_test_doc.pdf)

**Reply by JorjMcKie** [reply]
Thanks for the file original. Did a test - without being able to reproduce 😒. Here is the script and the resulting PDF - which is read by Adobe w/o complaints.
[test.zip](https://github.com/pymupdf/PyMuPDF/files/12385389/test.zip)

**Comment by djoltes** [comment]
That's odd; the files are fine in Chrome or Foxit, but not Adobe. I'm 
not in front of the code right now, but I've been doing intermediate 
saves as pages are processed and I'm saving with both deflate = True and 
garbage = 3. I'm wondering if that might be having an effect since I see 
an '/incremental=True' /option for ez_save(), but I can check on Monday.

Any ideas appreciated.

Thx,

dj

**Reply by JorjMcKie** [reply, marked as answer]
Does Adobe on your machine complain about my "x.pdf"?
Intermediate saves are not recommended, they are nothing but a waste of resources and like pulling the handbrake while driving.
Neither is incrmental save, because this leaves the body of the old file intact, i.e. all info is still there (and thus extractable by an expert), that you actually wanted to protect.

When redacting you should make sure to use garbage=3 at least (which is done in ez_save).

**Reply by djoltes** [reply]
Yes, your version viewed correctly and once I changed from incremental per-page save to a single ez_save() call at the end the Adobe problem vanished. Thanks, and be sure to add another pint to my 'tab' for later!
